openedx/core/djangoapps/programs/tasks/v1/tasks.py

```diff
@@ -0,0 +1,220 @@
+    # feature, this task should not have been invoked in the first place, and
+    # an error somewhere is likely (though a race condition is also possible).
+    # In either case, the task should not be executed nor should it be retried.
+    if not ProgramsApiConfig.current().is_certification_enabled:
+        LOGGER.warning(
+            'Task award_program_certificates cannot be executed when program certification is disabled in API config',
+        )
+        return
+
+    if not CredentialsApiConfig.current().is_learner_issuance_enabled:
+        LOGGER.warning(
+            'Task award_program_certificates cannot be executed when credentials issuance is disabled in API config',
+        )
+        return
+
+    try:
+        try:
+            student = User.objects.get(username=username)
+        except User.DoesNotExist:
+            LOGGER.exception('Task award_program_certificates was called with invalid username %s', username)
+            # Don't retry for this case - just conclude the task.
+            return
+
+        # Fetch the set of all course runs for which the user has earned a
+        # certificate.
+        course_certs = get_completed_courses(student)
+        if not course_certs:
+            # Highly unlikely, since at present the only trigger for this task
+            # is the earning of a new course certificate.  However, it could be
+            # that the transaction in which a course certificate was awarded
+            # was subsequently rolled back, which could lead to an empty result
+            # here, so we'll at least log that this happened before exiting.
+            #
+            # If this task is ever updated to support revocation of program
+            # certs, this branch should be removed, since it could make sense
+            # in that case to call this task for a user without any (valid)
+            # course certs.
+            LOGGER.warning('Task award_program_certificates was called for user %s with no completed courses', username)
+            return
+
+        # Invoke the Programs API completion check endpoint to identify any
+        # programs that are satisfied by these course completions.
+        programs_client = get_api_client(ProgramsApiConfig.current(), student)
+        program_ids = get_completed_programs(programs_client, course_certs)
+        if not program_ids:
+            # Again, no reason to continue beyond this point unless/until this
+            # task gets updated to support revocation of program certs.
+            return
+
+        # Determine which program certificates the user has already been
+        # awarded, if any.
+        existing_program_ids = get_awarded_certificate_programs(student)
+
+    except Exception, exc:  # pylint: disable=broad-except
+        LOGGER.exception('Failed to determine program certificates to be awarded for user %s', username)
+        raise self.retry(exc=exc)
```

Comment:
@rlucioni are you comfortable with implementing whatever we decide for custom retry in a follow-up PR?  I'd like to get the core functionality merged ASAP so that integration-related work isn't blocked.